Readme.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Prometheus components for Golang

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The module contains components for working with meters in the Prometheus service. The PrometheusCounters and PrometheusMetricsService components allow you to work both in client mode through PushGateway, and as a service.

The module contains the following packages:
- **Build** - the default factories for constructing components.
- **Count** - components of counters (metrics) with sending data to Prometheus via PushGateway
- **Services** - components of the service for reading counters (metrics) by the Prometheus service

<a name="links"></a> Quick links:

* [Configuration](http://docs.pipservices.org/conceptual/configuration/component_configuration/)
* [API Reference](https://godoc.org/github.com/pip-services3-gox/pip-services3-prometheus-gox)
* [Change Log](CHANGELOG.md)
* [Get Help](http://docs.pipservices.org/get_help/)
* [Contribute](http://docs.pipservices.org/contribute/)

## Use

Get the package from the Github repository:
```bash
go get -u github.com/pip-services3-gox/pip-services3-prometheus-gox@latest
```

## Develop

For development you shall install the following prerequisites:
* Golang v1.18+
* Visual Studio Code or another IDE of your choice
* Docker
* Git

Run automated tests:
```bash
go test -v ./test/...
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized test as:
```bash
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov** and **Levichev Dmitry**.

The documentation is written by:
- **Levichev Dmitry**

CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Prometheus components for Golang Changelog

## <a name="1.0.0"></a> 1.0.0 (2022-07-10)

Initial public release

### Features
* **build** the default factories for constructing components
* **count** components of counters (metrics) with sending data to Prometheus
* **services** components of the service for reading counters (metrics)

test/fixture/CountersFixture.go
package test_fixture

import (
	"context"
	"testing"
	"time"

	ccount "github.com/pip-services3-gox/pip-services3-components-gox/count"
	"github.com/stretchr/testify/assert"
)

type CountersFixture struct {
	counters *ccount.CachedCounters
}

func NewCountersFixture(counters *ccount.CachedCounters) *CountersFixture {
	cf := CountersFixture{}
	cf.counters = counters
	return &cf
}

func (c *CountersFixture) TestSimpleCounters(t *testing.T) {
	c.counters.Last(context.Background(), "Test.LastValue", 123)
	c.counters.Last(context.Background(), "Test.LastValue", 123456)

	var counter, _ = c.counters.Get(context.Background(), "Test.LastValue", ccount.LastValue)
	assert.NotNil(t, counter)
	assert.NotNil(t, counter.Last())
	assert.Equal(t, counter.Last(), float64(123456), 3)

	c.counters.IncrementOne(context.Background(), "Test.Increment")
	c.counters.Increment(context.Background(), "Test.Increment", 3)

	counter, _ = c.counters.Get(context.Background(), "Test.Increment", ccount.Increment)
	assert.NotNil(t, counter)
	assert.Equal(t, counter.Count(), int64(4))

	c.counters.TimestampNow(context.Background(), "Test.Timestamp")
	c.counters.TimestampNow(context.Background(), "Test.Timestamp")

	counter, _ = c.counters.Get(context.Background(), "Test.Timestamp", ccount.Timestamp)
	assert.NotNil(t, counter)
	assert.NotNil(t, counter.Time())

	c.counters.Stats(context.Background(), "Test.Statistics", 1)
	c.counters.Stats(context.Background(), "Test.Statistics", 2)
	c.counters.Stats(context.Background(), "Test.Statistics", 3)

	counter, _ = c.counters.Get(context.Background(), "Test.Statistics", ccount.Statistics)
	assert.NotNil(t, counter)
	assert.Equal(t, counter.Average(), float64(2), 3)

	c.counters.Dump(context.Background())

	select {
	case <-time.After(1000 * time.Millisecond):
		{
		}
	}
}

func (c *CountersFixture) TestMeasureElapsedTime(t *testing.T) {
	timer := c.counters.BeginTiming(context.Background(), "Test.Elapsed")

	time.AfterFunc(100*time.Millisecond, func() {
		timer.EndTiming(context.Background())

		counter, _ := c.counters.Get(context.Background(), "Test.Elapsed", ccount.Interval)
		assert.True(t, counter.Last() > 50)
		assert.True(t, counter.Last() < 5000)

		c.counters.Dump(context.Background())
		select {
		case <-time.After(1000 * time.Millisecond):
			{
			}
		}
	})
}


test/count/PrometheusCounters_test.go
package test_count

import (
	"context"
	"os"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	pcount "github.com/pip-services3-gox/pip-services3-prometheus-gox/count"
	pfixture "github.com/pip-services3-gox/pip-services3-prometheus-gox/test/fixture"
)

func TestPrometheusCounters(t *testing.T) {
	var counters *pcount.PrometheusCounters
	var fixture *pfixture.CountersFixture

	host := os.Getenv("PUSHGATEWAY_SERVICE_HOST")
	if host == "" {
		host = "localhost"
	}
	port := os.Getenv("PUSHGATEWAY_SERVICE_PORT")
	if port == "" {
		port = "9091"
	}
	counters = pcount.NewPrometheusCounters()
	fixture = pfixture.NewCountersFixture(counters.CachedCounters)

	config := cconf.NewConfigParamsFromTuples(
		"source", "test",
		"connection.host", host,
		"connection.port", port,
		"connection.protocol", "http",
	)
	counters.Configure(context.Background(), config)

	counters.Open(context.Background(), "")

	defer counters.Close(context.Background(), "")

	t.Run("Simple Counters", fixture.TestSimpleCounters)
	t.Run("Measure Elapsed Time", fixture.TestMeasureElapsedTime)
}


test/services/PrometheusMetricsService_test.go
package test_services

import (
	"context"
	"io/ioutil"
	"net/http"
	"testing"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	cinfo "github.com/pip-services3-gox/pip-services3-components-gox/info"
	pcount "github.com/pip-services3-gox/pip-services3-prometheus-gox/count"
	pservice "github.com/pip-services3-gox/pip-services3-prometheus-gox/services"
	"github.com/stretchr/testify/assert"
)

func TestPrometheusMetricsService(t *testing.T) {
	ctx := context.Background()
	var service *pservice.PrometheusMetricsService
	var counters *pcount.PrometheusCounters

	var restConfig = cconf.NewConfigParamsFromTuples(
		"connection.protocol", "http",
		"connection.host", "localhost",
		"connection.port", "3000",
	)

	service = pservice.NewPrometheusMetricsService()
	service.Configure(ctx, restConfig)

	counters = pcount.NewPrometheusCounters()

	contextInfo := cinfo.NewContextInfo()
	contextInfo.Name = "Test"
	contextInfo.Description = "This is a test container"

	references := cref.NewReferencesFromTuples(ctx,
		cref.NewDescriptor("pip-services", "context-info", "default", "default", "1.0"), contextInfo,
		cref.NewDescriptor("pip-services", "counters", "prometheus", "default", "1.0"), counters,
		cref.NewDescriptor("pip-services", "metrics-service", "prometheus", "default", "1.0"), service,
	)
	counters.SetReferences(ctx, references)
	service.SetReferences(ctx, references)

	opnErr := counters.Open(ctx, "")
	if opnErr == nil {
		service.Open(ctx, "")
	}

	defer service.Close(ctx, "")
	defer counters.Close(ctx, "")

	var url = "http://localhost:3000"

	counters.IncrementOne(ctx, "test.counter1")
	counters.Stats(ctx, "test.counter2", 2)
	counters.Last(ctx, "test.counter3", 3)
	counters.TimestampNow(ctx, "test.counter4")

	getRes, getErr := http.Get(url + "/metrics")
	assert.Nil(t, getErr)
	assert.NotNil(t, getRes)
	assert.True(t, getRes.StatusCode < 400)
	body, _ := ioutil.ReadAll(getRes.Body)
	assert.True(t, len(body) > 0)
}


count/PrometheusCounters.go
package count

import (
	"bytes"
	"context"
	"net/http"
	"os"
	"sync"
	"time"

	cconf "github.com/pip-services3-gox/pip-services3-commons-gox/config"
	cerr "github.com/pip-services3-gox/pip-services3-commons-gox/errors"
	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	ccount "github.com/pip-services3-gox/pip-services3-components-gox/count"
	cinfo "github.com/pip-services3-gox/pip-services3-components-gox/info"
	clog "github.com/pip-services3-gox/pip-services3-components-gox/log"
	rpcconnect "github.com/pip-services3-gox/pip-services3-rpc-gox/connect"
)

// PrometheusCounters performance counters that send their metrics to Prometheus service.
//
// The component is normally used in passive mode conjunction with PrometheusMetricsService.
// Alternatively when connection parameters are set it can push metrics to Prometheus PushGateway.
//
//	Configuration parameters:
//
//		- connection(s):
//			- discovery_key:         (optional) a key to retrieve the connection from connect.idiscovery.html IDiscovery
//			- protocol:              connection protocol: http or https
//			- host:                  host name or IP address
//			- port:                  port number
//			- uri:                   resource URI or connection string with all parameters in it
//		- options:
//			- retries:               number of retries (default: 3)
//			- connect_timeout:       connection timeout in milliseconds (default: 10 sec)
//			- timeout:               invocation timeout in milliseconds (default: 10 sec)
//
//	References:
//
//		- *:logger:*:*:1.0         (optional) ILogger components to pass log messages
//		- *:counters:*:*:1.0         (optional) ICounters components to pass collected measurements
//		- *:discovery:*:*:1.0        (optional)  IDiscovery services to resolve connection
//
// See:  RestService
// See:  CommandableHttpService
//
// Example:
//		ctx := context.Background()
//		counters = NewPrometheusCounters();
//		counters.Configure(ctx, cconf.NewConfigParamsFromTuples(
//		    "connection.protocol", "http",
//		    "connection.host", "localhost",
//		    "connection.port", 8080
//		));
//
//		counters.Open("123")
//
//		counters.Increment(ctx, "mycomponent.mymethod.calls");
//		timing := counters.BeginTiming(ctx, "mycomponent.mymethod.exec_time");
//		    ...
//		timing.EndTiming(ctx);
//
//		counters.Dump(ctx);
//
type PrometheusCounters struct {
	*ccount.CachedCounters
	logger             *clog.CompositeLogger
	connectionResolver *rpcconnect.HttpConnectionResolver
	opened             bool
	source             string
	instance           string
	client             *http.Client
	requestRoute       string
	timeout            int
	retries            int
	connectTimeout     int
	uri                string

	Lock sync.Mutex
}

// NewPrometheusCounters is creates a new instance of the performance counters.
// Returns *PrometheusCounters
// pointer on new instance
func NewPrometheusCounters() *PrometheusCounters {
	c := PrometheusCounters{}
	c.CachedCounters = ccount.InheritCacheCounters(&c)
	c.logger = clog.NewCompositeLogger()
	c.connectionResolver = rpcconnect.NewHttpConnectionResolver()
	c.opened = false
	c.timeout = 10000
	c.retries = 3
	c.connectTimeout = 10000
	return &c
}

// Configure method are configures component by passing configuration parameters.
//	Parameters:
//		- ctx context.Context	operation context
//		- config   *cconf.ConfigParams
// configuration parameters to be set.
func (c *PrometheusCounters) Configure(ctx context.Context, config *cconf.ConfigParams) {
	c.CachedCounters.Configure(ctx, config)
	c.connectionResolver.Configure(ctx, config)

	c.source = config.GetAsStringWithDefault("source", c.source)
	c.instance = config.GetAsStringWithDefault("instance", c.instance)
	c.retries = config.GetAsIntegerWithDefault("options.retries", c.retries)
	c.connectTimeout = config.GetAsIntegerWithDefault("options.connectTimeout", c.connectTimeout)
	c.timeout = config.GetAsIntegerWithDefault("options.timeout", c.timeout)
}

// SetReferences method are sets references to dependent components.
//	Parameters:
//		- ctx context.Context	operation context
//		- references  cref.IReferences
// references to locate the component dependencies.
func (c *PrometheusCounters) SetReferences(ctx context.Context, references cref.IReferences) {
	c.logger.SetReferences(ctx, references)
	c.connectionResolver.SetReferences(ctx, references)
	ref := references.GetOneOptional(
		cref.NewDescriptor("pip-services", "context-info", "default", "*", "1.0"))
	contextInfo, _ := ref.(*cinfo.ContextInfo)
	if contextInfo != nil && c.source == "" {
		c.source = contextInfo.Name
	}
	if contextInfo != nil && c.instance == "" {
		c.instance = contextInfo.ContextId
	}
}

// IsOpen method are checks if the component is opened.
// Returns true if the component has been opened and false otherwise.
func (c *PrometheusCounters) IsOpen() bool {
	return c.opened
}

// Open method are opens the component.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId 	string (optional) transaction id to trace execution through call chain.
// Returns error
//	error or nil, if no errors occured.
func (c *PrometheusCounters) Open(ctx context.Context, correlationId string) (err error) {
	if c.opened {
		return nil
	}

	c.opened = true
	connection, _, err := c.connectionResolver.Resolve(correlationId)

	c.Lock.Lock()
	if err != nil {
		c.client = nil
		c.Lock.Unlock()
		c.logger.Warn(ctx, correlationId, "Connection to Prometheus server is not configured: "+err.Error())
		return nil
	}

	c.Lock.Unlock()
	c.uri = connection.Uri()

	job := c.source
	if job == "" {
		job = "unknown"
	}

	instance := c.instance
	if instance == "" {
		host, _ := os.Hostname()
		instance = host
	}
	c.requestRoute = "/metrics/job/" + job + "/instance/" + instance

	localClient := http.Client{}
	localClient.Timeout = (time.Duration)(c.timeout) * time.Millisecond

	c.Lock.Lock()
	defer c.Lock.Unlock()
	c.client = &localClient
	if c.client == nil {
		ex := cerr.NewConnectionError(correlationId, "CANNOT_CONNECT", "Connection to REST service failed").WithDetails("url", c.uri)
		return ex
	}

	return nil
}

// Close method are closes component and frees used resources.
//	Parameters:
//		- ctx context.Context	operation context
//		- correlationId string (optional) transaction id to trace execution through call chain.
// Returns error
// error or nil, if no errors occured.
func (c *PrometheusCounters) Close(ctx context.Context, correlationId string) error {
	c.opened = false

	c.Lock.Lock()
	c.client = nil
	c.Lock.Unlock()

	c.requestRoute = ""
	return nil
}

// Save method are saves the current counters measurements.
//	Parameters:
//		- ctx context.Context	operation context
//		- counters   []ccount.Counter current counters measurements to be saves.
// Retruns error
// error or nil, if no errors occured.
func (c *PrometheusCounters) Save(cxt context.Context, counters []ccount.Counter) (err error) {
	c.Lock.Lock()
	if c.client == nil {
		c.Lock.Unlock()
		return nil
	}
	c.Lock.Unlock()

	url := c.uri + c.requestRoute

	body := PrometheusCounterConverter.ToString(counters, "", "")

	req, reqErr := http.NewRequest(http.MethodPut, url, bytes.NewBuffer([]byte(body)))
	if reqErr != nil {
		err = cerr.NewUnknownError("PrometheusCounters", "UNSUPPORTED_METHOD", "Method is not supported by REST client").WithDetails("verb", "PUT").WithCause(reqErr)
		return err
	}
	// Set headers
	req.Header.Set("Accept", "text/html")
	retries := c.retries
	var resp *http.Response
	var respErr error

	for retries > 0 {
		// Try send request
		resp, respErr = c.client.Do(req)
		if respErr != nil {

			retries--
			if retries == 0 {
				err = cerr.NewUnknownError("PrometheusCounters", "COMMUNICATION_ERROR", "Unknown communication problem on REST client").WithCause(respErr)
				return err
			}
			continue

		}
		break
	}
	if resp != nil {
		defer resp.Body.Close()
		if resp.StatusCode >= 204 && resp.StatusCode < 300 {
			return nil
		}
		c.logger.Error(req.Context(), "prometheus-counters", respErr, "Failed to push metrics to prometheus")
	}

	return respErr
}


count/PrometheusCounterConverter.go
package count

import (
	"strings"

	cconv "github.com/pip-services3-gox/pip-services3-commons-gox/convert"
	ccount "github.com/pip-services3-gox/pip-services3-components-gox/count"
)

//  PrometheusCounterConverter is helper class that converts performance counter values into
//  a response from Prometheus metrics service.
var PrometheusCounterConverter TPrometheusCounterConverter = TPrometheusCounterConverter{}

type TPrometheusCounterConverter struct {
}

// ToString method converts the given counters to a string that is returned by Prometheus metrics service.
//	Parameters:
//		- counters  a list of counters to convert.
//		- source    a source (context) name.
//		- instance  a unique instance name (usually a host name).
// Returns string
// string view of counter
func (c *TPrometheusCounterConverter) ToString(counters []ccount.Counter, source string, instance string) string {

	if len(counters) == 0 {
		return ""
	}

	var builder string = ""

	for _, counter := range counters {
		counterName := c.parseCounterName(counter)
		labels := c.generateCounterLabel(counter, source, instance)

		switch counter.Type {
		case ccount.Increment:
			builder += "# TYPE " + counterName + " gauge\n"
			builder += counterName + labels + " " + cconv.StringConverter.ToString(counter.Count) + "\n"
		case ccount.Interval:
			builder += "# TYPE " + counterName + "_max gauge\n"
			builder += counterName + "_max" + labels + " " + cconv.StringConverter.ToString(counter.Max) + "\n"
			builder += "# TYPE " + counterName + "_min gauge\n"
			builder += counterName + "_min" + labels + " " + cconv.StringConverter.ToString(counter.Min) + "\n"
			builder += "# TYPE " + counterName + "_average gauge\n"
			builder += counterName + "_average" + labels + " " + cconv.StringConverter.ToString(counter.Average) + "\n"
			builder += "# TYPE " + counterName + "_count gauge\n"
			builder += counterName + "_count" + labels + " " + cconv.StringConverter.ToString(counter.Count) + "\n"
		case ccount.LastValue:
			builder += "# TYPE " + counterName + " gauge\n"
			builder += counterName + labels + " " + cconv.StringConverter.ToString(counter.Last) + "\n"
		case ccount.Statistics:
			builder += "# TYPE " + counterName + "_max gauge\n"
			builder += counterName + "_max" + labels + " " + cconv.StringConverter.ToString(counter.Max) + "\n"
			builder += "# TYPE " + counterName + "_min gauge\n"
			builder += counterName + "_min" + labels + " " + cconv.StringConverter.ToString(counter.Min) + "\n"
			builder += "# TYPE " + counterName + "_average gauge\n"
			builder += counterName + "_average" + labels + " " + cconv.StringConverter.ToString(counter.Average) + "\n"
			builder += "# TYPE " + counterName + "_count gauge\n"
			builder += counterName + "_count" + labels + " " + cconv.StringConverter.ToString(counter.Count) + "\n"
		case ccount.Timestamp: // Prometheus doesn't support non-numeric metrics
			builder += "# TYPE " + counterName + " gauge\n" //" untyped\n"
			builder += counterName + labels + " " + cconv.StringConverter.ToString(counter.Time.Unix()) + "\n"
		}
	}

	return builder
}

func (c *TPrometheusCounterConverter) AtomicCountersToCounters(atomicCounters []*ccount.AtomicCounter) []ccount.Counter {
	counters := make([]ccount.Counter, len(atomicCounters))

	for _, atomicCounter := range atomicCounters {
		counter := ccount.Counter{
			Name:    atomicCounter.Name(),
			Type:    atomicCounter.Type(),
			Last:    atomicCounter.Last(),
			Count:   atomicCounter.Count(),
			Min:     atomicCounter.Min(),
			Max:     atomicCounter.Max(),
			Average: atomicCounter.Average(),
			Time:    atomicCounter.Time(),
		}

		counters = append(counters, counter)
	}

	return counters

}

func (c *TPrometheusCounterConverter) generateCounterLabel(counter ccount.Counter, source string, instance string) string {

	labels := make(map[string]string, 0)

	if source != "" {
		labels["source"] = source
	}

	if instance != "" {
		labels["instance"] = instance
	}

	nameParts := strings.Split(counter.Name, ".")

	// If there are other predictable names from which we can parse labels, we can add them below
	if len(nameParts) >= 3 && nameParts[2] == "exec_time" {
		labels["service"] = nameParts[0]
		labels["command"] = nameParts[1]
	}

	if len(labels) == 0 {
		return ""
	}

	builder := "{"
	for key := range labels {
		if len(builder) > 1 {
			builder += ","
		}
		builder += key + `="` + labels[key] + `"`
	}
	builder += "}"

	return builder
}

func (c *TPrometheusCounterConverter) parseCounterName(counter ccount.Counter) string {
	if counter.Name == "" {
		return ""
	}

	nameParts := strings.Split(counter.Name, ".")

	// If there are other predictable names from which we can parse labels, we can add them below
	if len(nameParts) >= 3 && nameParts[2] == "exec_time" {
		return nameParts[2]
	}

	// TODO: are there other assumptions we can make?
	// Or just return as a single, valid name
	result := strings.ToLower(counter.Name)
	result = strings.Replace(result, ".", "_", -1)
	result = strings.Replace(result, "/", "_", -1)

	return result
}

func (c *TPrometheusCounterConverter) parseCounterLabels(counter ccount.Counter, source string, instance string) interface{} {
	labels := make(map[string]string, 0)

	if source != "" {
		labels["source"] = source
	}

	if instance != "" {
		labels["instance"] = instance
	}

	nameParts := strings.Split(counter.Name, ".")

	// If there are other predictable names from which we can parse labels, we can add them below
	if len(nameParts) >= 3 && nameParts[2] == "exec_time" {
		labels["service"] = nameParts[0]
		labels["command"] = nameParts[1]
	}

	return labels
}


services/PrometheusMetricsService.go
package services

import (
	"context"
	"io"
	"net/http"

	cref "github.com/pip-services3-gox/pip-services3-commons-gox/refer"
	ccount "github.com/pip-services3-gox/pip-services3-components-gox/count"
	cinfo "github.com/pip-services3-gox/pip-services3-components-gox/info"
	pcount "github.com/pip-services3-gox/pip-services3-prometheus-gox/count"
	rpcservices "github.com/pip-services3-gox/pip-services3-rpc-gox/services"
)

// PrometheusMetricsService is service that exposes "/metrics" route for Prometheus to scap performance metrics.
//
//	Configuration parameters:
//
//		- dependencies:
//			- endpoint:              override for HTTP Endpoint dependency
//			- prometheus-counters:   override for PrometheusCounters dependency
//		- connection(s):
//			- discovery_key:         (optional) a key to retrieve the connection from IDiscovery
//			- protocol:              connection protocol: http or https
//			- host:                  host name or IP address
//			- port:                  port number
//			- uri:                   resource URI or connection string with all parameters in it
//
//	References:
//
//		- *:logger:*:*:1.0         (optional)  ILogger components to pass log messages
//		- *:counters:*:*:1.0         (optional)  ICounters components to pass collected measurements
//		- *:discovery:*:*:1.0        (optional)  IDiscovery services to resolve connection
//		- *:endpoint:http:*:1.0          (optional)  HttpEndpoint reference to expose REST operation
//		- *:counters:prometheus:*:1.0    PrometheusCounters reference to retrieve collected metrics
//
// See RestService
// See RestClient
//
// Example
//		ctx := context.Context()
//		service := NewPrometheusMetricsService();
//		service.Configure(ctx, cconf.NewConfigParamsFromTuples(
//		    "connection.protocol", "http",
//		    "connection.host", "localhost",
//		    "connection.port", "8080",
//		));
//
//		err := service.Open(ctx, "123")
//		if  err == nil {
//		    fmt.Println("The Prometheus metrics service is accessible at http://localhost:8080/metrics");
//		    defer service.Close(ctx, "")
//		}
//
type PrometheusMetricsService struct {
	rpcservices.RestService
	cachedCounters *ccount.CachedCounters
	source         string
	instance       string
}

// NewPrometheusMetricsService are creates a new instance of c service.
// Returns *PrometheusMetricsService
// pointer on new instance
func NewPrometheusMetricsService() *PrometheusMetricsService {
	c := &PrometheusMetricsService{}
	c.RestService = *rpcservices.InheritRestService(c)
	c.DependencyResolver.Put(context.Background(), "cached-counters", cref.NewDescriptor("pip-services", "counters", "cached", "*", "1.0"))
	c.DependencyResolver.Put(context.Background(), "prometheus-counters", cref.NewDescriptor("pip-services", "counters", "prometheus", "*", "1.0"))
	return c
}

// SetReferences is sets references to dependent components.
//	Parameters:
//		- ctx context.Context	operation context
//		- references cref.IReferences
// references to locate the component dependencies.
func (c *PrometheusMetricsService) SetReferences(ctx context.Context, references cref.IReferences) {
	c.RestService.SetReferences(ctx, references)

	resolv := c.DependencyResolver.GetOneOptional("prometheus-counters")
	c.cachedCounters = resolv.(*pcount.PrometheusCounters).CachedCounters
	if c.cachedCounters == nil {
		resolv = c.DependencyResolver.GetOneOptional("cached-counters")
		c.cachedCounters = resolv.(*ccount.CachedCounters)
	}
	ref := references.GetOneOptional(
		cref.NewDescriptor("pip-services", "context-info", "default", "*", "1.0"))
	contextInfo := ref.(*cinfo.ContextInfo)

	if contextInfo != nil && c.source == "" {
		c.source = contextInfo.Name
	}
	if contextInfo != nil && c.instance == "" {
		c.instance = contextInfo.ContextId
	}
}

// Register method are registers all service routes in HTTP endpoint.
func (c *PrometheusMetricsService) Register() {
	c.RegisterRoute("get", "metrics", nil, func(res http.ResponseWriter, req *http.Request) { c.metrics(res, req) })
}

// Handles metrics requests
//	Parameters:
//		- req   an HTTP request
//		- res   an HTTP response
func (c *PrometheusMetricsService) metrics(res http.ResponseWriter, req *http.Request) {

	var atomicCounters []*ccount.AtomicCounter
	if c.cachedCounters != nil {
		atomicCounters = c.cachedCounters.GetAll()
	}

	counters := pcount.PrometheusCounterConverter.AtomicCountersToCounters(atomicCounters)
	body := pcount.PrometheusCounterConverter.ToString(counters, c.source, c.instance)

	res.Header().Add("content-type", "text/plain")
	res.WriteHeader(200)
	_, wrErr := io.WriteString(res, (string)(body))
	if wrErr != nil {
		c.Logger.Error(req.Context(), "PrometheusMetricsService", wrErr, "Can't write response")
	}
}
